### ISSUE TITLE
`M3-3D-Complete` > Paste each block as a separate GitHub issue in `atom-ahg-plugins`

### ISSUE THREAD

**Issue by johanpiet2** [ArchiveHeritageGroup/atom-extensions-catalog#229]
# GitHub Issues — Milestone: `M3-3D-Complete`
> Paste each block as a separate GitHub issue in `atom-ahg-plugins`

---

## Issue 1 — Bug
**Title:** `Three.js loaded from CDN at r128 (2021) — pin and serve locally`

**Labels:** `bug` `3d` `csp` `priority:high`

**Description:**
Three.js is loaded from CDN at version r128 (November 2021). Current version is r170+. This creates three problems:

1. CDN dependency — violates CSP policy (same issue as OSD/Mirador)
2. r128 is too old to use ES module imports correctly — caused the `importmap` workaround needed in December
3. OBJLoader, STLLoader, OrbitControls are loaded as separate CDN scripts — these are now part of the Three.js `examples/jsm/` module tree and the CDN paths have changed

**Current broken CDN references:**
```
https://cdnjs.cloudflare.com/ajax/libs/three.js/r128/three.min.js
https://cdn.jsdelivr.net/npm/three@0.128.0/examples/js/loaders/OBJLoader.js
https://cdn.jsdelivr.net/npm/three@0.128.0/examples/js/loaders/STLLoader.js
https://cdn.jsdelivr.net/npm/three@0.128.0/examples/js/controls/OrbitControls.js
```

**Fix required:**
- Download Three.js r128 (or upgrade to r160+) to `ahgThemeB5Plugin/vendor/threejs/`
- Bundle OBJLoader, STLLoader, OrbitControls alongside
- Update all template references to local paths
- Remove CDN references from Gallery plugin templates

**Files to update:**
- `ahgThemeB5Plugin/lib/helper/IiifViewerHelper.php`
- `ahgGalleryPlugin/modules/ahgGalleryPlugin/templates/indexSuccess.php`
- `ahgGalleryPlugin/modules/ahgGalleryPlugin/templates/editSuccess.php`

---

## Issue 2 — Bug
**Title:** `Google Model Viewer loaded from Google CDN — serve locally for CSP and offline compliance`

**Labels:** `bug` `3d` `csp` `priority:high`

**Description:**
Google Model Viewer is loaded from Google's CDN:
```
https://ajax.googleapis.com/ajax/libs/model-viewer/3.3.0/model-viewer.min.js
```

Same CSP and availability problem as Three.js. Additionally, institutions with strict network policies (government archives, classified environments) cannot call out to Google APIs.

**Fix required:**
- Download `model-viewer.min.js` v3.3.0 to `ahgThemeB5Plugin/vendor/model-viewer/`
- Update all template references

**Files to update:**
- `ahgThemeB5Plugin` 3D viewer templates
- `ahgGalleryPlugin` templates
- `iiif-3d-manifest.php`

---

## Issue 3 — Bug
**Title:** `GLB MIME type not confirmed in nginx — silent 3D loading failure`

**Labels:** `bug` `3d` `nginx` `priority:high`

**Description:**
GLB files require correct MIME type in nginx for model-viewer and Three.js GLTFLoader to load them. If not set, the file serves as `application/octet-stream` and model-viewer silently fails.

**Required nginx mime.types entry:**
```nginx
model/gltf-binary    glb;
model/gltf+json      gltf;
```

**Verify on server:**
```bash
grep -i "glb\|gltf\|model" /etc/nginx/mime.types
```

**Fix required:**
- Confirm entries exist in `/etc/nginx/mime.types`
- Add if missing and reload nginx
- Add to framework install documentation

---

## Issue 4 — Bug
**Title:** `FBX format listed as supported but FBXLoader not included in Three.js stack`

**Labels:** `bug` `3d` `priority:medium`

**Description:**
FBX is listed as a supported 3D format in documentation and `extension.json` but `FBXLoader` requires:
- `fflate` (compression library)
- Separate loader not present in the current r128 Three.js setup

FBX files will silently fail to load — no error shown to user.

**Options:**
1. Remove FBX from supported formats list until implemented
2. Add `FBXLoader` + `fflate` dependency and implement properly

**Recommendation:** Option 1 for now — remove from supported list, add back when properly implemented. FBX is a proprietary Autodesk format; GLB/GLTF is the better target format for GLAM use.

**Files to update:**
- `ahgThemeB5Plugin` format detection list
- `ahgGalleryPlugin` format detection
- Any `extension.json` listing supported formats
- User documentation

---

## Issue 5 — Enhancement
**Title:** `3D hotspot placement UI — click to place annotation on model surface`

**Labels:** `enhancement` `3d` `annotations` `priority:medium`

**Description:**
The database schema for 3D hotspots is complete (`object_3d_hotspot`, `object_3d_hotspot_i18n`) and the service layer (`Model3DService::addHotspot()`) exists, but there is no front-end UI to actually place a hotspot on a model.

**What exists:**
- `object_3d_hotspot` table — stores `position_x/y/z` and `normal_x/y/z`
- `Model3DService::addHotspot()` — API to save hotspot data
- Admin `executeHotspots` — list/manage existing hotspots
- Google Model Viewer native hotspot slot support

**What's missing:**
- Edit mode in the viewer — "Add hotspot" toggle
- Click handler on model surface to capture `position` and `normal` from raycaster
- Form to enter hotspot title, description, type, color
- Save to `Model3DService::addHotspot()`

**Implementation approach for Google Model Viewer:**
```javascript
// model-viewer exposes surface hit test
viewer.addEventListener('click', (e) => {
    const hit = viewer.surfaceFromPoint(e.clientX, e.clientY);
    if (hit) {
        // hit.position = {x, y, z} in model space
        // hit.normal = {x, y, z}
        openHotspotForm(hit.position, hit.normal);
    }
});
```

**For Three.js (OBJ/STL):**
```javascript
// Raycaster on click
raycaster.setFromCamera(mouse, camera);
const intersects = raycaster.intersectObjects(scene.children, true);
if (intersects.length > 0) {
    const point = intersects[0].point;
    const normal = intersects[0].face.normal;
    openHotspotForm(point, normal);
}
```

**Hotspot types to support:**
- `annotation` — General note
- `info` — Informational point
- `damage` — Condition/damage marker (links to Condition Assessment)
- `detail` — Point of interest
- `link` — External URL

---

## Issue 6 — Enhancement
**Title:** `3D hotspot rendering in Three.js viewer — only implemented for model-viewer`

**Labels:** `enhancement` `3d` `annotations` `priority:medium`

**Description:**
3D hotspots are only rendered in the Google Model Viewer (which supports native `<button slot="hotspot">` elements). The Three.js viewer (used for OBJ/STL) has no hotspot overlay rendering.

**Fix required:**
- Load hotspots from API for the current `digital_object_id`
- Project 3D coordinates to screen using Three.js camera
- Render HTML overlay buttons at projected positions
- Update positions on camera move via `OrbitControls` event

```javascript
// Project world position to screen
function projectToScreen(position, camera, renderer) {
    const vector = position.clone().project(camera);
    const x = (vector.x * 0.5 + 0.5) * renderer.domElement.clientWidth;
    const y = (-(vector.y * 0.5) + 0.5) * renderer.domElement.clientHeight;
    return { x, y, visible: vector.z < 1 };
}
```

**Dependencies:** Issue 5 (hotspot placement UI)

---

## Issue 7 — Enhancement
**Title:** `IIIF 3D manifest — wire iiif-3d-manifest.php into main manifest service`

**Labels:** `enhancement` `3d` `iiif` `priority:medium`

**Description:**
`iiif-3d-manifest.php` exists as a standalone root-level file but is not:
- Wired into the main `IiifManifestService`
- Accessible via the standard `/iiif/manifest/{slug}` route
- Returning a properly structured IIIF 3.0 manifest with `rendering` property for 3D

**Current state:** Standalone PHP file, not integrated.

**Required:**
The IIIF community approach for 3D is to include the 3D model as a `rendering` property on the canvas, with a standard image thumbnail as the canvas body:

```json
{
  "id": "https://example.com/iiif/manifest/my-3d-object",
  "type": "Manifest",
  "label": { "en": ["Roman Vase"] },
  "items": [{
    "id": "https://example.com/iiif/canvas/1",
    "type": "Canvas",
    "items": [{ 
      "type": "AnnotationPage",
      "items": [{
        "type": "Annotation",
        "motivation": "painting",
        "body": { "id": "thumbnail.jpg", "type": "Image" },
        "target": "canvas/1"
      }]
    }],
    "rendering": [{
      "id": "https://example.com/uploads/.../model.glb",
      "type": "Model",
      "format": "model/gltf-binary",
      "label": { "en": ["3D Model"] }
    }]
  }]
}
```

**Fix required:**
- Move 3D manifest logic into `IiifManifestService::generateManifest()` — detect 3D format and add `rendering` property
- Route `/iiif/manifest/{slug}` handles 3D automatically
- Deprecate standalone `iiif-3d-manifest.php`

---

## Issue 8 — Enhancement
**Title:** `Consider O3DV as replacement for Three.js OBJ/STL viewer`

**Labels:** `enhancement` `3d` `priority:low`

**Description:**
[Online 3D Viewer (O3DV)](https://github.com/kovacsv/Online3DViewer) is a purpose-built 3D model viewer library specifically designed for the "show this model" use case — which is exactly what GLAM institutions need.

**Advantages over raw Three.js:**
- Supports 30+ formats out of the box including OBJ, STL, FBX, PLY, DAE, 3DS, STEP
- No need to manually load format-specific loaders
- Built-in camera controls, lighting, background
- Smaller integration surface — less code to maintain
- MIT licensed
- Actively maintained

**Relevance for Heratio:**
- Solves the FBX problem (Issue 4) without manual FBXLoader setup
- Reduces Three.js maintenance burden
- Better format coverage for museum 3D objects (STEP for engineering/technical objects)

**Proposed approach:**
- Use `model-viewer` for GLB/GLTF (keep current — it's best for this)
- Replace Three.js OBJ/STL with O3DV
- O3DV handles everything else (FBX, PLY, DAE, STEP, 3DS)

**This is an evaluation issue** — test O3DV against current Three.js implementation before committing.

---

## Issue 9 — Enhancement
**Title:** `Link 3D hotspot damage markers to Condition Assessment plugin`

**Labels:** `enhancement` `3d` `condition` `museum` `priority:low`

**Description:**
The `object_3d_hotspot` table has a `hotspot_type` of `damage`. This should link to the Condition Assessment plugin (`arConditionPlugin`) so that:

1. A conservator marks a damage area on a 3D model
2. The hotspot creates a linked condition assessment record
3. The condition record appears in the `arConditionPlugin` workflow
4. The 3D viewer shows damage hotspots colour-coded by severity

**This is a museum-specific feature** — relevant for Spectrum 5.0 condition checking workflows.

**Dependencies:**
- Issue 5 (hotspot placement UI)
- Issue 6 (hotspot rendering in Three.js)
- `arConditionPlugin` must be enabled

---

## Suggested Milestone Structure

```
M3-3D-Complete
├── Fix before any demo (bugs)
│   ├── Issue 1 — Three.js local vendor
│   ├── Issue 2 — model-viewer local vendor
│   └── Issue 3 — GLB MIME type nginx
├── Accuracy (format support)
│   └── Issue 4 — Remove FBX from supported list
├── Core 3D annotation features
│   ├── Issue 5 — Hotspot placement UI ← build first
│   ├── Issue 6 — Hotspot rendering Three.js
│   └── Issue 7 — IIIF 3D manifest integration
└── Strategic (evaluate/future)
    ├── Issue 8 — Evaluate O3DV
    └── Issue 9 — Damage hotspot → Condition Assessment link
```

**Comment by johanpiet2** [comment]
implemented
